File: lib/src/main/java/ir/arefdev/irdebitcardscanner/ResourceModelFactory.java
package ir.arefdev.irdebitcardscanner;

import android.content.Context;
import android.content.res.AssetFileDescriptor;

import java.io.FileInputStream;
import java.io.IOException;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;

class ResourceModelFactory {

	private static ResourceModelFactory instance;

	static ResourceModelFactory getInstance() {
		if (instance == null) {
			instance = new ResourceModelFactory();
		}

		return instance;
	}

	MappedByteBuffer loadFindFourFile(Context context) throws IOException {
		return loadModelFromResource(context, R.raw.findfour);
	}

	MappedByteBuffer loadRecognizeDigitsFile(Context context) throws IOException {
		return loadModelFromResource(context, R.raw.fourrecognize);
	}

	private MappedByteBuffer loadModelFromResource(Context context, int resource) throws IOException {
		AssetFileDescriptor fileDescriptor = context.getResources().openRawResourceFd(resource);
		FileInputStream inputStream = new FileInputStream(fileDescriptor.getFileDescriptor());
		FileChannel fileChannel = inputStream.getChannel();
		long startOffset = fileDescriptor.getStartOffset();
		long declaredLength = fileDescriptor.getDeclaredLength();
		MappedByteBuffer result = fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength);
		inputStream.close();
		fileDescriptor.close();
		return result;
	}
}

File: lib/src/main/java/ir/arefdev/irdebitcardscanner/ResourceModelFactory.kt
package ir.arefdev.irdebitcardscanner

import android.content.Context
import android.content.res.AssetFileDescriptor
import com.arefbhrn.irdebitcardscanner.R
import java.io.FileInputStream
import java.io.IOException
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel

internal object ResourceModelFactory {

    @Throws(IOException::class)
    fun loadFindFourFile(context: Context): MappedByteBuffer =
        loadModelFromResource(context, R.raw.findfour)

    @Throws(IOException::class)
    fun loadRecognizeDigitsFile(context: Context): MappedByteBuffer =
        loadModelFromResource(context, R.raw.fourrecognize)

    @Throws(IOException::class)
    private fun loadModelFromResource(context: Context, resource: Int): MappedByteBuffer {
        val fileDescriptor: AssetFileDescriptor = context.resources.openRawResourceFd(resource)
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel: FileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declaredLength = fileDescriptor.declaredLength
        val result = fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
        inputStream.close()
        fileDescriptor.close()
        return result
    }
}
